File: internal/deployment/modules_test.go
package deployment

import (
	"encoding/json"
	"strings"
	"testing"

	"github.com/helmrdotdev/helmr/internal/jsoncanon"
)

func TestModuleMapMatchesSharedGoldenFixture(t *testing.T) {
	fixture := loadContractFixture(t)
	moduleMap, err := ParseModuleMap([]byte(fixture.ModuleMap.Canonical))
	if err != nil {
		t.Fatal(err)
	}
	canonical, err := CanonicalModuleMap(moduleMap)
	if err != nil {
		t.Fatal(err)
	}
	if string(canonical) != fixture.ModuleMap.Canonical {
		t.Fatalf("canonical module map = %q, want %q", canonical, fixture.ModuleMap.Canonical)
	}
	if moduleMap.Modules[2].Path != "packages/shared/src/\ue000.ts" || moduleMap.Modules[3].Path != "packages/shared/src/😀.ts" {
		t.Fatalf("module map does not use unsigned UTF-8 path order: %q, %q", moduleMap.Modules[2].Path, moduleMap.Modules[3].Path)
	}
}

func TestModuleMapRejectsSharedMutations(t *testing.T) {
	fixture := loadContractFixture(t)
	for _, test := range fixture.ModuleMapRejections {
		t.Run(test.Name, func(t *testing.T) {
			var value map[string]any
			if err := json.Unmarshal([]byte(fixture.ModuleMap.Canonical), &value); err != nil {
				t.Fatal(err)
			}
			modules := value["modules"].([]any)
			first := modules[0].(map[string]any)
			switch test.Mutation {
			case "missing_format_version":
				delete(value, "formatVersion")
			case "unknown_root_member":
				value["unknown"] = true
			case "transformer":
				value["transformer"] = "helmr.typescript.v1"
			case "module_order":
				modules[0], modules[1] = modules[1], modules[0]
			case "duplicate_path":
				modules[1] = first
			case "absolute_path":
				setFixtureModulePath(first, "/packages/shared/src/legacy.cts")
			case "escaping_path":
				setFixtureModulePath(first, "packages/../src/legacy.cts")
			case "backslash_path":
				setFixtureModulePath(first, `packages\shared\src\legacy.cts`)
			case "reserved_helmr_root":
				setFixtureModulePath(first, "helmr/legacy.cts")
			case "reserved_dot_helmr_root":
				setFixtureModulePath(first, ".helmr/legacy.cts")
			case "reserved_node_modules_root":
				setFixtureModulePath(first, "node_modules/legacy.cts")
			case "declaration_path":
				setFixtureModulePath(first, "packages/shared/src/legacy.d.cts")
			case "unsupported_extension":
				setFixtureModulePath(first, "packages/shared/src/legacy.tsx")
			case "source_digest":
				first["sourceDigest"] = "sha256:invalid"
			case "code_digest":
				first["codeDigest"] = "sha256:invalid"
			case "format":
				first["format"] = "module"
				setFixtureModulePath(first, first["path"].(string))
			case "code_path_key":
				first["codePath"] = "helmr/files/modules/" + strings.Repeat("0", 64) + ".cjs"
			case "code_path_extension":
				first["codePath"] = strings.TrimSuffix(first["codePath"].(string), ".cjs") + ".mjs"
			case "unknown_module_member":
				first["unknown"] = true
			default:
				t.Fatalf("unknown fixture mutation %q", test.Mutation)
			}
			raw, err := json.Marshal(value)
			if err != nil {
				t.Fatal(err)
			}
			canonical, err := jsoncanon.Transform(raw)
			if err != nil {
				t.Fatal(err)
			}
			if _, err := ParseModuleMap(canonical); err == nil {
				t.Fatal("ParseModuleMap returned nil error")
			}
		})
	}
}

func setFixtureModulePath(module map[string]any, path string) {
	module["path"] = path
	module["codePath"] = moduleCodePath(path, ModuleFormat(module["format"].(string)))
}

func TestModuleMapAcceptsEmptyModuleArray(t *testing.T) {
	moduleMap := ModuleMap{
		FormatVersion: ModuleMapFormatVersion,
		Modules:       []Module{},
		Transformer:   TypeScriptTransformer,
	}
	canonical, err := CanonicalModuleMap(moduleMap)
	if err != nil {
		t.Fatal(err)
	}
	if _, err := ParseModuleMap(canonical); err != nil {
		t.Fatal(err)
	}
}

func TestModuleMapRejectsNonCanonicalAndBoundedInput(t *testing.T) {
	fixture := loadContractFixture(t)
	if _, err := ParseModuleMap([]byte(" " + fixture.ModuleMap.Canonical)); err == nil {
		t.Fatal("ParseModuleMap accepted non-canonical input")
	}
	if _, err := ParseModuleMap(nil); err == nil {
		t.Fatal("ParseModuleMap accepted empty input")
	}
	if _, err := ParseModuleMap(make([]byte, maxProgramFileSizeBytes+1)); err == nil {
		t.Fatal("ParseModuleMap accepted oversized input")
	}
}

func TestModuleMapRejectsTooManyModules(t *testing.T) {
	moduleMap := ModuleMap{
		FormatVersion: ModuleMapFormatVersion,
		Modules:       make([]Module, maxModuleCount+1),
		Transformer:   TypeScriptTransformer,
	}
	if err := ValidateModuleMap(moduleMap); err == nil {
		t.Fatal("ValidateModuleMap accepted too many modules")
	}
}

File: sdk/typescript/src/internal/modules.test.ts
import { describe, expect, test } from "bun:test"
import { createHash } from "node:crypto"
import { readFile } from "node:fs/promises"
import { resolve } from "node:path"

import { canonicalizeJson } from "./jsoncanon"
import {
  canonicalModuleMap,
  MODULE_MAP_FORMAT_VERSION,
  parseModuleMap,
  TYPESCRIPT_TRANSFORMER,
  validateModuleMap,
  type ModuleMap,
} from "./modules"

interface ModuleMapFixture {
  readonly moduleMap: { readonly canonical: string }
  readonly moduleMapRejections: readonly {
    readonly name: string
    readonly mutation: string
  }[]
}

const decoder = new TextDecoder()

describe("deployment module map", async () => {
  const fixture = await loadFixture()

  test("matches the shared canonical fixture and unsigned UTF-8 path order", () => {
    const moduleMap = parseModuleMap(fixture.moduleMap.canonical)
    expect(decoder.decode(canonicalModuleMap(moduleMap))).toBe(fixture.moduleMap.canonical)
    expect(moduleMap.modules.slice(2).map((module) => module.path)).toEqual([
      "packages/shared/src/\ue000.ts",
      "packages/shared/src/😀.ts",
    ])
  })

  test("rejects the shared invalid mutations", () => {
    for (const item of fixture.moduleMapRejections) {
      const value = JSON.parse(fixture.moduleMap.canonical) as Record<string, unknown>
      const modules = value["modules"] as Array<Record<string, unknown>>
      const first = modules[0] as Record<string, unknown>
      switch (item.mutation) {
        case "missing_format_version":
          delete value["formatVersion"]
          break
        case "unknown_root_member":
          value["unknown"] = true
          break
        case "transformer":
          value["transformer"] = "helmr.typescript.v1"
          break
        case "module_order":
          ;[modules[0], modules[1]] = [modules[1] as Record<string, unknown>, modules[0] as Record<string, unknown>]
          break
        case "duplicate_path":
          modules[1] = { ...first }
          break
        case "absolute_path":
          setFixtureModulePath(first, "/packages/shared/src/legacy.cts")
          break
        case "escaping_path":
          setFixtureModulePath(first, "packages/../src/legacy.cts")
          break
        case "backslash_path":
          setFixtureModulePath(first, "packages\\shared\\src\\legacy.cts")
          break
        case "reserved_helmr_root":
          setFixtureModulePath(first, "helmr/legacy.cts")
          break
        case "reserved_dot_helmr_root":
          setFixtureModulePath(first, ".helmr/legacy.cts")
          break
        case "reserved_node_modules_root":
          setFixtureModulePath(first, "node_modules/legacy.cts")
          break
        case "declaration_path":
          setFixtureModulePath(first, "packages/shared/src/legacy.d.cts")
          break
        case "unsupported_extension":
          setFixtureModulePath(first, "packages/shared/src/legacy.tsx")
          break
        case "source_digest":
          first["sourceDigest"] = "sha256:invalid"
          break
        case "code_digest":
          first["codeDigest"] = "sha256:invalid"
          break
        case "format":
          first["format"] = "module"
          setFixtureModulePath(first, first["path"] as string)
          break
        case "code_path_key":
          first["codePath"] = `helmr/files/modules/${"0".repeat(64)}.cjs`
          break
        case "code_path_extension":
          first["codePath"] = (first["codePath"] as string).replace(/\.cjs$/, ".mjs")
          break
        case "unknown_module_member":
          first["unknown"] = true
          break
        default:
          throw new Error(`unknown fixture mutation ${item.mutation}`)
      }
      const canonical = canonicalizeJson(JSON.stringify(value))
      expect(() => parseModuleMap(canonical), item.name).toThrow()
    }
  })

  test("accepts an empty module array for an all-JavaScript program", () => {
    const moduleMap: ModuleMap = {
      formatVersion: MODULE_MAP_FORMAT_VERSION,
      modules: [],
      transformer: TYPESCRIPT_TRANSFORMER,
    }
    expect(() => parseModuleMap(canonicalModuleMap(moduleMap))).not.toThrow()
  })

  test("rejects non-canonical, empty, and oversized input", () => {
    expect(() => parseModuleMap(` ${fixture.moduleMap.canonical}`)).toThrow(/canonical/)
    expect(() => parseModuleMap(new Uint8Array())).toThrow(/size/)
    expect(() => parseModuleMap(new Uint8Array(16777217))).toThrow(/size/)
  })

  test("rejects a string containing an unpaired surrogate instead of repairing it", () => {
    const path = "packages/shared/src/\ufffd.ts"
    const valid: ModuleMap = {
      formatVersion: MODULE_MAP_FORMAT_VERSION,
      modules: [
        {
          codeDigest: `sha256:${"a".repeat(64)}`,
          codePath: fixtureModuleCodePath(path, "module"),
          format: "module",
          path,
          sourceDigest: `sha256:${"1".repeat(64)}`,
        },
      ],
      transformer: TYPESCRIPT_TRANSFORMER,
    }
    const repairedCanonical = decoder.decode(canonicalModuleMap(valid))
    expect(() => parseModuleMap(repairedCanonical.replace("\ufffd", "\ud800"))).toThrow(/surrogate/)
  })

  test("rejects more than 65,536 entries", () => {
    const moduleMap = {
      formatVersion: MODULE_MAP_FORMAT_VERSION,
      modules: Array.from({ length: 65537 }, () => null),
      transformer: TYPESCRIPT_TRANSFORMER,
    } as unknown as ModuleMap
    expect(() => validateModuleMap(moduleMap)).toThrow(/maximum/)
  })
})

async function loadFixture(): Promise<ModuleMapFixture> {
  const path = resolve(import.meta.dir, "../../../../fixtures/contracts/deployment-v0/golden.json")
  return JSON.parse(await readFile(path, "utf8")) as ModuleMapFixture
}

function setFixtureModulePath(module: Record<string, unknown>, path: string): void {
  module["path"] = path
  module["codePath"] = fixtureModuleCodePath(path, module["format"] as string)
}

function fixtureModuleCodePath(path: string, format: string): string {
  const key = createHash("sha256")
    .update("helmr.typescript-module.v0")
    .update(new Uint8Array([0]))
    .update(path)
    .digest("hex")
  return `helmr/files/modules/${key}.${format === "module" ? "mjs" : "cjs"}`
}
